### PR TITLE
Added TCP transport for use with `speculos`

### PR DESCRIPTION
**Pull request by ryankurte**
a simple tokio based TCP transport impl, for use with `speculos`'s `--apdu-port`.

(also wondering whether y'all are open to other changes / improvements before i dig too far into this?)

:link: [zboto Link](https://app.clickup.com/t/2u483x3)

### CODE REVIEW

File: ledger-transport-tcp/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/// Speculos (TCP) transport errors
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("IO error {:?}", 0)]
+    Io(std::io::Error),
```
Comment (becominginsane) [inline ledger-transport-tcp/src/error.rs:5]: wouldn't this be the same as connection? what other types of IO do we do here?

File: ledger-transport-tcp/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+pub use error::Error;
+
+/// Ledger speculos (TCP) transport
+pub struct TransportTcp {
+    s: Mutex<TcpStream>,
+    timeout: Duration,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "clap", derive(clap::Parser))]
```
Comment (becominginsane) [inline ledger-transport-tcp/src/lib.rs:26]: I guess this is to make it easier to be used in cli applications by having a subcommand? I think it's unnecessary/overkill, and considering it's such a limited number of options I'd rather have it as arguments to the `new` function...

This also allows us to drop the `humantime` dependency and just have any `Duration` as argument.

As for the defaults, either have options for the arguments and document it, or use a builder-like pattern, which imo is more idiomatic

**Review by becominginsane — CHANGES_REQUESTED** [review]
First of all, thank you for your contribution! This is similar to what I had in mind so great job!

I have left some comments that I think will make the crate leaner and more idiomatic :)

File: ledger-transport-tcp/src/lib.rs (hunk 2 above)
Comment (ryankurte) [reply ledger-transport-tcp/src/lib.rs:26]: > I guess this is to make it easier to be used in cli applications by having a subcommand?
> considering it's such a limited number of options I'd rather have it as arguments to the new function...

sortof, it makes it easier to embed configuration / provide defaults, and simplifies using the library / helps unify the configuration across different consumers. it's an easy addition and if it's not provided here it has to be implemented by every user of the library.

`clap` is the ecosystem standard for this (now derive has moved from `structopt`), i can switch it from default to opt in?

> As for the defaults, either have options for the arguments and document it, or use a builder-like pattern, which imo is more idiomatic

implementing `Default::default` gives you the same functionality without actually having a builder (which i have now done), and the configuration object is easy to wrap or drop into `clap` (or `serde`, though i haven't needed this -yet-)

Comment (becominginsane) [reply ledger-transport-tcp/src/lib.rs:27]: > it has to be implemented by every user of the library
yes that would be the point, but I find it unlikely that the struct would be used directly anyways, unless it's a very minimal cli... 
I guess removing it from the default set of features is sufficient

**Review by becominginsane — APPROVED** [review]

**Pull request closed without merging**